### PR TITLE
Receive and verify AIK key

### PR DESCRIPTION
**Pull request by arturkow2000**

### CODE REVIEW

File: pal/pal_pc/src/trussed/store.rs [hunk 1]
```diff
@@ -63,9 +63,11 @@ impl littlefs2::driver::Storage for Flash {
     const READ_SIZE: usize = 4;
     const WRITE_SIZE: usize = 4;
 
-    // For now let's use 64 KiB only, may be extended when needed.
+    // 64 KiB is not enough - LittleFS wastes much of available storage
+    // see https://github.com/littlefs-project/littlefs/issues/645
+    // increase size to 128 KiB
     const BLOCK_SIZE: usize = 4096;
```
Comment (krystian-hebel) [inline pal/pal_pc/src/trussed/store.rs:69]: Can we make it smaller in order to waste less space per directory?

File: src/certmgr/verify.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                    // FIXME:
+                    // We cannot call self.verify_internal() while iterating because
+                    // both iter_certificate() and verify_internal() need exclusive
+                    // access to trussed.
+                    //
+                    // For now we load certificates into memory before checking them.
+                    // While this works when there is a small number of certificates
+                    // it will fill memory when there are more.
+                    let potential_parents: Vec<X509Certificate> =
+                        self.iter_certificates(organization, trussed).collect();
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]: Is it possible to return vector of file names (or paths if they are in different directories) instead and load full file contents inside `for` loop below?

File: src/client/mod.rs [hunk 3]
```diff
@@ -79,37 +85,142 @@ impl<'a> FobnailClient<'a> {
...
+                            ek_cert,
+                        )
+                    },
+                ) {
+                    Ok((secret, id_object, encrypted_secret)) => {
+                        // Trick to move data without copying.
+                        // Rust won't allow us just move out data because doing
+                        // so would make state invalid.
+                        let mut aik = vec![];
+                        core::mem::swap(data, &mut aik);
```
Comment (krystian-hebel) [inline src/client/mod.rs:166]: I don't know about Rust's memory management, but shouldn't this be swapped back to initial `data` so it is properly freed?

File: src/client/mod.rs [hunk 4]
```diff
@@ -79,37 +85,142 @@ impl<'a> FobnailClient<'a> {
...
+                        id_object: &id_object,
+                        encrypted_secret: &encrypted_secret,
+                    })
+                    .unwrap();
+
+                    let mut request = coap_lite::CoapRequest::new();
+                    request.set_path("/challenge");
+                    request.set_method(RequestType::Post);
+                    request.message.payload = encoded.to_vec();
+                    // TODO: should set Content-Type header to CBOR
```
Comment (krystian-hebel) [inline src/client/mod.rs]:
```suggestion
                    request.message.set_content_format(coap_lite::ContentFormat::ApplicationCBOR);
```

This isn't checked on Attester side right now, but at some point we may depend on it.

File: src/client/mod.rs (hunk 3 above)
Comment (krystian-hebel) [reply src/client/mod.rs:166]: OK, I notices it is done in `handle_server_response()`. Maybe mention this in comments, here and in `handle_server_response()`?

File: src/client/state.rs [hunk 5]
```diff
@@ -12,20 +14,59 @@ pub enum State<'a> {
...
+        aik: Vec<u8>,
+    },
+
+    /// Parse and load AIK key.
+    LoadAik { raw_aik: Vec<u8> },
+
     // TODO: before requesting metadata we must request AIK (Attestation
     // Identity Key) key, receive it and verify.
     // This adds two new states: RequestAik and VerifyAik inserted between
     // InitDataReceived and RequestMetadata
```
Comment (krystian-hebel) [inline src/client/state.rs]: Already done
```suggestion
```

File: src/client/tpm/kdf.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+pub fn kdf_a(
+    algo: Algorithm,
+    key: &[u8],
+    label: &str,
+    context_u: &[u8],
+    context_v: &[u8],
+    bits: u32,
+) -> Vec<u8> {
+    // FIXME: Trussed has some support for KDF, but we cannot use it because of
+    // lack of RSA support (and possibly for other reasons two).
```
Comment (krystian-hebel) [inline src/client/tpm/kdf.rs]:
```suggestion
    // lack of RSA support (and possibly for other reasons too).
```

File: src/client/tpm/mod.rs [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+                // See section 24.4 of TPM 2.0 specification, part 1.
+                let aes_key = kdf::kdf_a(
+                    loaded_key_name.algorithm(),
+                    &seed,
+                    "STORAGE",
+                    loaded_key_name.raw_data(),
+                    &[],
+                    (seed.len() * 8).try_into().unwrap(),
+                );
+                // panic!("AES key = {:#x?}", aes_key);
```
Comment (krystian-hebel) [inline src/client/tpm/mod.rs]: Remove

File: src/client/tpm/mu.rs [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+                // let symmetric_key_bits = decode!(cursor, data, u16);
+                // let symmetric_sym_mode = decode!(cursor, data, u16);
+
+                let rsa_scheme = Algorithm::try_from_u8(decode!(cursor, data, u16))?;
+                let rsa_scheme_details = decode!(cursor, data, u16);
+                // TODO: do we need to support this?
+                ensure!(
+                    rsa_scheme.is_null() && rsa_scheme_details == 0x10,
+                    "Non-null scheme, this is not supported"
+                );
```
Comment (krystian-hebel) [inline src/client/tpm/mu.rs]: No, it is present, but it is `TPM2_ALG_NULL`. In that case its other fields (`keyBits`, `mode`, `details`) do not exist. This is easy to miss in TPM specification, it is hidden somewhere behind the definition of `+` symbol either prepended or appended to type name. Same is true for `scheme` - it is also `TPM2_ALG_NULL` so it has no `details`. In the current code what should be `symmetric.algorithm` is read as `scheme.scheme`, and what should be `scheme.scheme` is read as `scheme.detail`.

```suggestion
                let symmetric_algorithm_id = Algorithm::try_from_u8(decode!(cursor, data, u16))?;
                ensure!(
                    symmetric_algorithm_id.is_null(),
                    "Non-null symmetric algorithm, this is not supported"
                );
                
                let rsa_scheme = Algorithm::try_from_u8(decode!(cursor, data, u16))?;
                ensure!(
                    rsa_scheme.is_null(),
                    "Non-null scheme, this is not supported"
                );
```

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: pal/pal_pc/src/trussed/store.rs (hunk 1 above)
Comment (krystian-hebel) [reply pal/pal_pc/src/trussed/store.rs:69]: No, we can't, this is flash erase granularity.

File: src/certmgr/verify.rs (hunk 2 above)
Comment (arturkow2000) [reply src/certmgr/verify.rs]: @krystian-hebel I modified `iter_certificates()` a bit so we can now call verify_internal while iterating.

File: src/client/mod.rs (hunk 4 above)
Comment (arturkow2000) [reply src/client/mod.rs]: Fixed

File: src/client/tpm/mu.rs (hunk 8 above)
Comment (arturkow2000) [reply src/client/tpm/mu.rs]: Fixed

File: src/client/mod.rs (hunk 3 above)
Comment (arturkow2000) [reply src/client/mod.rs:166]: @krystian-hebel I mentioned it.

> I don't know about Rust's memory management, but shouldn't this be swapped back to initial data so it is properly freed?

No. When `aik` goes out of scope, Vec descrutctor will be called and data will freed normally.

**Comment by macpijan** [comment]
@arturkow2000 @krystian-hebel What's left to close this one?

**Comment by arturkow2000** [comment]
@krystian-hebel Anything left?

File: src/client/tpm/mu.rs [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    inner: &'a [u8],
+}
+
+impl<'a> ByteArray<'a> {
+    pub fn new(buffer: &'a [u8]) -> Self {
+        Self { inner: buffer }
+    }
+
+    pub fn encode(&self) -> Vec<u8> {
+        let buf_len: u16 = self.inner.len().try_into().expect("buffer to big for TPM");
```
Comment (krystian-hebel) [inline src/client/tpm/mu.rs]:
```suggestion
        let buf_len: u16 = self.inner.len().try_into().expect("buffer too big for TPM");
```

**Review by krystian-hebel — COMMENTED** [review]
One typo, otherwise LGTM

File: src/client/tpm/mu.rs (hunk 9 above)
Comment (arturkow2000) [reply src/client/tpm/mu.rs]: Fixed

**Review by krystian-hebel — APPROVED** [review]

**Comment by tomzy-0** [comment]
Merged

**Pull request merged**